### PR TITLE
🌱 Improve e2e structure

### PR DESCRIPTION
**Pull request by mboukhalfa**
Run `upgrade` in a separate test spec and get rid of `if else` that was replaced by ginkgo `focus` and `skip` to select the tests.
Preparing to run `remediation` and `features` tests in parallel by also moving `remediation` test to separate test spec.
The parallelism will be implemented by running `remediation` and `features` tests in different vms from jenkins or project infra else we can have a phony test trigger for each

### CODE REVIEW

**Comment by lentzi90** [comment]
Nice job! It is a bit hard to follow with so large changes, but from what I can tell this looks good.

**Comment by lentzi90** [comment]
Nice job! It is a bit hard to follow with so large changes, but from what I can tell this looks good.

**Review by lentzi90 — COMMENTED** [review]
Nice job! It is a bit hard to follow with so large changes, but from what I can tell this looks good.

**Review by fmuyassarov — COMMENTED** [review]
/lgtm

**Review by smoshiur1237 — COMMENTED** [review]
Thanks nice job. It looks good but  I think it needs to clarify how to add new feature and also it would be nice to change the feature test file name. Also, this process will add double infra for each e2e trigger(v1b1, v1a5, ephemeral test) and need to confirm how our CI will react with that extra resource consumption.

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -1,97 +0,0 @@
...
-					upgradeBMO(bootstrapClusterProxy.GetClientSet())
-					upgradeIronic(bootstrapClusterProxy.GetClientSet())
-					certRotation(bootstrapClusterProxy.GetClientSet(), bootstrapClusterProxy.GetClient())
-					nodeReuse(bootstrapClusterProxy.GetClient())
-
-				}
-			}
-		})
-	})
-})
```
Comment (furkatgofurov7) [inline test/e2e/e2e_test.go:97]: Dumping this file completely and renaming it to "features_test.go" does not look the right direction to take in order to put parallelism in place. I will list all the main concerns I have and we have discussed during the call offline:

- We are shifting from CAPI structured e2e test file names and conventions for file names across the providers. One can argue this, but I prefer sticking to common test file names for a better user experience
- In the current state, we are leaving remediation tests alone (by the way this test also incorporates inspection tests in the middle) outside and planning to have them separate for purposes of running them in parallel. However, that way we are kind of saying that remediation+inspection tests are not the features but whatever tests we have in a new features_test.go are. 
- How are you planning to keep this scalable, i,e in the future? **how does**/**based on what** the one would decide where to add/run the new feature e2e test (in the features_test.go OR separately as you have remediation test)?

Personally, I had imagined the parallelism as a bit different approach, where we could run all the tests in the same VM under the hood, in one go. Since that is not the case (we would have 2 separate VMs running different tests at the same time), I question myself what is the benefit of it over a new job/trigger you could introduce, which after all both of them would consume the same resources. 
If needed, we could classify and gate our current e2e tests, so we have different different triggers based on that classification, as [CAPI](https://github.com/kubernetes-sigs/cluster-api/pull/6411) or [CAPZ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2235#issuecomment-1101038608) for example does:

- test-v1b1-ubuntu-e2e-main
- test-v1b1-ubuntu-e2e-optional
- test-v1b1-ubuntu-capi-e2e

File: docs/e2e-test.md [hunk 2]
```diff
@@ -29,24 +49,43 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
 
 ## Included tests
 
-The e2e tests currently run one after the other, using the same environment for all of them.
-They run in this order:
+The e2e tests currently include two different sets:
+
+`features_test`: split into two test specs
```
Comment (furkatgofurov7) [inline docs/e2e-test.md:54]: this looks redundant, maybe we should say something along these lines:
 
```
 The e2e tests currently include three different sets:
1. pivoting_based_feature_tests
2. remediation_feature_tests
3. upgrade_tests
```
?

File: docs/e2e-test.md [hunk 3]
```diff
@@ -29,24 +49,43 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
...
 The ephemeral cluster is first launched using [metal3-dev-env](https://github.com/metal3-io/metal3-dev-env).
-The remediation and inspection tests are then run with the controllers still in the ephemeral cluster.
-After this, the controllers are moved to the target cluster during pivoting.
-From this point on all tests are run in the target cluster, until the controllers are moved back to the ephemeral cluster during re-pivoting.
+The remediation and inspection tests are then run with the controllers still in the ephemeral cluster either before pivoting or after re-pivoting.
+
+`upgrade_test`:
+
+- Upgrade BMO
+- Upgrade Ironic
```
Comment (furkatgofurov7) [inline docs/e2e-test.md:78]: are you missing the actual management cluster upgrade test?

File: docs/e2e-test.md [hunk 4]
```diff
@@ -29,24 +49,43 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
...
+
+Guidelines to follow when adding new E2E tests:
+
+- Tests should be defined in a new file and separate test spec, Unless the new test depends on an existing tests.
```
Comment (furkatgofurov7) [inline docs/e2e-test.md:82]:
```suggestion
- Tests should be defined in a new file and separate test spec, unless the new test depends on existing tests.
```

File: docs/e2e-test.md [hunk 5]
```diff
@@ -29,24 +49,43 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
...
+- Tests are categorized with a custom label that can be used to filter a set of E2E tests to be run.To run a subset of tests, a combination of either one or both of GINKGO_FOCUS and GINKGO_SKIP env variables can be set. the label are defined in the description of the `Describe` container between `[]` for example:
```
Comment (furkatgofurov7) [inline docs/e2e-test.md:83]:
```suggestion
- Tests are categorized with a custom label that can be used to filter a set of E2E tests to be run. To run a subset of tests, a combination of either one or both of the `GINKGO_FOCUS` and `GINKGO_SKIP` env variables can be set. The labels are defined in the description of the `Describe` container between `[]`, for example:
```

File: docs/e2e-test.md [hunk 6]
```diff
@@ -29,24 +49,43 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
...
+
+`[upgrade]` => Tests which verify upgrades including capi/capim3 upgrades, ironic/bmo upgrades..
+`[remediation]` => test which verify nodes remediation.
```
Comment (furkatgofurov7) [inline docs/e2e-test.md:86]:
```suggestion
`[upgrade]` => runs only existing upgrade tests including CAPI, CAPM3, Ironic and Baremetal Operator.```
`[remediation]` => runs only remediation and inspection tests
```

File: docs/e2e-test.md [hunk 7]
```diff
@@ -29,24 +49,43 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
...
+
+For instance, to skip the upgrade E2E tests set `GINKGO_SKIP="[upgrade]"`
```
Comment (furkatgofurov7) [inline docs/e2e-test.md:88]: can you make it clear, where this needs to be set, in Makefile?

**Review by furkatgofurov7 — COMMENTED** [review]
Thanks, looks good, last few comments to improve the documentation part:

**Review by furkatgofurov7 — APPROVED** [review]
/approve

**Review by smoshiur1237 — APPROVED** [review]
/lgtm
